### DISCUSSION TITLE
Added: LDM Notes for July 27th, 2020

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#3741] [LDM Notes]
https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-07-27.md

Agenda:

- [Improved nullable analysis in constructors](https://github.com/RikkiGibson/csharplang/blob/nullable-ctor/proposals/nullable-constructor-analysis.md)
- [Equality operators (`==` and `!=`) in records](https://github.com/dotnet/csharplang/issues/3707#issuecomment-661800278)
- `.ToString()` or `GetDebuggerDisplay()` on records?
- Restore W-warning to `T t = default;` for generic `T`, now you can write `T?`?

**Comment by huoyaoyuan** [comment]
> Improved nullable analysis in constructors

Happy to see that change. I've ran into this:
```csharp
foreach (var (key, value) in parameters)
{
    if (key == "name")
        Name = value;
}

if (Name is null) throw new Exception("No name in parameters!");
```
I had to assign the value to a variable inside loop, and assign it to the property after null check in the non-branch route.

> Equality operators (`==` and `!=`) in records

In practice, a big advantage of `==` operators is taking care of `null`... Writing `EqualityComparer.Default<T>.Equals` is annoying.

> `ToString` or `DebuggerDisplay` on record types

Anonymous types use `{ Prop = value }` format. Please consider consistency.

> Restore W-warning to `T t = default;` for generic `T`, now you can write `T?`

I wonder the future with non-defaultable value types. It's quite a mess around `T?` now, and I don't want things to be worse then.
Can the `T!` syntax that's proposed in very early phase of NRT help?

**Comment by erik-kallen** [comment]
Nice to see that you are considering improvements to nullable analysis in constructors, but the problem I mentioned in #2830 seems to still be unadressed.

**Comment by MgSam** [comment]
Very glad to see the `==` question has a final decision.

**Comment by 333fred** [comment]
> Nice to see that you are considering improvements to nullable analysis in constructors, but the problem I mentioned in #2830 seems to still be unadressed.

We're looking at required properties to solve nullable initialization issues.

**Comment by erik-kallen** [comment]
@333fred Great to hear that you are working on it! I would really like to enable nullable in my web projects, but until we have a solution for that problem it is not worth it IMO.
